File: ts/src/startupSync.ts
/**
 * Startup data sync orchestration.
 *
 * Split from server.ts. Owns the retry/backoff scheduling, single-flight
 * locking, and the cache-clearing cascade triggered when sync writes new data.
 * Mirrors python/src/prts_mcp/startup_sync.py.
 */

import { join } from "node:path";
import { loadConfig } from "./config.js";
import {
  clearOperatorCaches,
} from "./data/operator.js";
import { clearEnemyCaches } from "./data/enemy.js";
import { clearStageCaches } from "./data/stage.js";
import { clearItemCaches } from "./data/item.js";
import { clearStageEnemyCaches } from "./data/stageEnemy.js";
import { clearSearchCaches } from "./data/search.js";
import { clearStoryCaches } from "./data/story.js";
import { syncImages } from "./data/imagesSync.js";
import { syncRelease, syncReleaseArchivePair } from "./data/sync.js";
import { archiveSpecForDataset, releaseSpecForDataset, GAMEDATA_EXCEL, GAMEDATA_LEVELS, STORY_ZH_CN } from "./data/datasets.js";

// ---------------------------------------------------------------------------
// Logging helper (shared with server.ts)
// ---------------------------------------------------------------------------

function log(level: "INFO" | "WARN" | "ERROR", msg: string): void {
  const ts = new Date().toISOString();
  process.stderr.write(`${ts} ${level} prts_mcp.server: ${msg}\n`);
}

// ---------------------------------------------------------------------------
// Constants + single-flight state
// ---------------------------------------------------------------------------

const SYNC_RETRY_DELAYS_MS = [30_000, 120_000, 600_000] as const;
const AUTO_SYNC_DEFAULT_INTERVAL_SECONDS = 3600;
const AUTO_SYNC_MIN_INTERVAL_SECONDS = 60;
const AUTO_SYNC_MAX_INTERVAL_SECONDS = 7 * 24 * 60 * 60;
const syncInFlight = new Set<string>();
let autoSyncStarted = false;
type SyncRunResult = "retry" | "done" | "skipped";

function shouldRetrySync(status: string): boolean {
  return status === "offline_fallback" || status === "no_data";
}

/**
 * Decide whether the gamedata pair sync warrants a dense retry (#102).
 *
 * A generation (commitSha) mismatch alone must NOT trigger dense retries
 * (30s/120s/600s): when both archives are up-to-date, divergent commitShas
 * are stale local metadata, not missing data, and the next periodic cycle
 * resolves it. Only offline_fallback / no_data retry immediately.
 */
export function gamedataPairNeedsRetry(
  excelStatus: string,
  levelsStatus: string,
): boolean {
  return shouldRetrySync(excelStatus) || shouldRetrySync(levelsStatus);
}

async function singleFlightSync(label: string, runSync: () => Promise<boolean>): Promise<SyncRunResult> {
  if (syncInFlight.has(label)) {
    log("INFO", `${label} sync is already running; skipping overlapping attempt.`);
    return "skipped";
  }
  syncInFlight.add(label);
  try {
    return await runSync() ? "retry" : "done";
  } finally {
    syncInFlight.delete(label);
  }
}

function scheduleSyncRetry(
  label: string,
  runSync: () => Promise<boolean>,
  attempt = 0,
): void {
  const delayMs = SYNC_RETRY_DELAYS_MS[attempt];
  if (delayMs === undefined) {
    log("WARN", `${label} sync still needs retry after ${SYNC_RETRY_DELAYS_MS.length} attempts; waiting for the next periodic cycle or process start.`);
    return;
  }

  const timer = setTimeout(() => {
    void singleFlightSync(label, runSync)
      .then((result) => {
        if (result === "skipped") scheduleSyncRetry(label, runSync, attempt);
        else if (result === "retry") scheduleSyncRetry(label, runSync, attempt + 1);
      })
      .catch((err: unknown) => {
        log("ERROR", `${label} retry sync threw unexpectedly: ${err instanceof Error ? err.message : String(err)}`);
        scheduleSyncRetry(label, runSync, attempt + 1);
      });
  }, delayMs);
  timer.unref();

  log("INFO", `${label} sync will retry in ${Math.round(delayMs / 1000)}s.`);
}

// ---------------------------------------------------------------------------
// Startup sync entry point
// ---------------------------------------------------------------------------

export async function runStartupSync(forceCheck = false): Promise<void> {
  const cfg = loadConfig();
  const startupTasks: Promise<void>[] = [];

  // Gamedata sync
  if (cfg.isCustomGamedata) {
    log("INFO", `GAMEDATA_PATH is custom (${cfg.gamedataPath}); auto-sync disabled.`);
  } else {
    const archiveSpec = archiveSpecForDataset(
      GAMEDATA_EXCEL,
      join(cfg.gamedataPath, "archives", "zh_CN-excel.zip"),
      cfg.gamedataPath,
    );
    const levelsSpec = archiveSpecForDataset(
      GAMEDATA_LEVELS,
      join(cfg.levelsPath, "archives", "zh_CN-levels.zip"),
      cfg.levelsPath,
    );

    const runGamedataSync = async (): Promise<boolean> => {
      const [r, levelsResult] = await syncReleaseArchivePair(
        archiveSpec,
        levelsSpec,
        forceCheck,
      );
      const sha = r.commitSha ? r.commitSha.slice(0, 8) : "unknown";
      if (r.status === "updated") {
        log("INFO", `Data updated from GitHub Release (${r.spec.repo} @ ${sha}).`);
      } else if (r.status === "up_to_date") {
        log("INFO", `Data is up to date (${r.spec.repo} @ ${sha}).`);
      } else if (r.status === "offline_fallback") {
        log("WARN", `Network unavailable; using cached data (${r.spec.repo} @ ${sha}). Error: ${r.error}`);
      } else {
        log("ERROR", `Sync failed for ${r.spec.repo} — no data. Error: ${r.error}`);
      }
      const levelsSha = levelsResult.commitSha
        ? levelsResult.commitSha.slice(0, 8)
        : "unknown";
      if (levelsResult.status === "updated") {
        log("INFO", `Level data updated from GitHub Release (${levelsResult.spec.repo} @ ${levelsSha}).`);
      } else if (levelsResult.status === "up_to_date") {
        log("INFO", `Level data is up to date (${levelsResult.spec.repo} @ ${levelsSha}).`);
      } else if (levelsResult.status === "offline_fallback") {
        log("WARN", `Network unavailable; using cached level data (${levelsResult.spec.repo} @ ${levelsSha}). Error: ${levelsResult.error}`);
      } else {
        log("ERROR", `Level data sync failed for ${levelsResult.spec.repo} — no data. Error: ${levelsResult.error}`);
      }
      const sameGeneration = r.commitSha !== null
        && r.commitSha === levelsResult.commitSha;
      if (
        sameGeneration
        && (r.status === "updated" || levelsResult.status === "updated")
      ) {
        clearOperatorCaches();
        clearEnemyCaches();
        clearStageCaches();
        clearItemCaches();
        clearStageEnemyCaches();
        clearSearchCaches();
      }
      return gamedataPairNeedsRetry(r.status, levelsResult.status);
    };

    startupTasks.push(
      singleFlightSync("Gamedata", runGamedataSync)
        .catch((err: unknown) => {
          log("ERROR", `Startup sync threw unexpectedly: ${err instanceof Error ? err.message : String(err)}`);
          return true;
        })
        .then((result) => {
          if (result !== "done" && !forceCheck) {
            scheduleSyncRetry("Gamedata", runGamedataSync);
          }
        }),
    );
  }

  // Storyjson release sync (always attempt unless user supplied STORYJSON_PATH)
  if (!process.env["STORYJSON_PATH"]) {
    const releaseSpec = releaseSpecForDataset(STORY_ZH_CN, cfg.storyjsonZip);

    const runStorySync = async (): Promise<boolean> => {
      const r = await syncRelease(releaseSpec, forceCheck);
      const sha = r.commitSha ? r.commitSha.slice(0, 8) : "unknown";
      if (r.status === "updated") {
        clearStoryCaches();
        log("INFO", `Storyjson updated from GitHub Release (${r.spec.repo} @ ${sha}).`);
      } else if (r.status === "up_to_date") {
        log("INFO", `Storyjson is up to date (${r.spec.repo} @ ${sha}).`);
      } else if (r.status === "offline_fallback") {
        log("WARN", `Network unavailable; using cached storyjson (${r.spec.repo} @ ${sha}). Error: ${r.error}`);
      } else {
        log("ERROR", `Storyjson sync failed for ${r.spec.repo} — no zip. Error: ${r.error}`);
      }
      return shouldRetrySync(r.status);
    };

    startupTasks.push(
      singleFlightSync("Storyjson", runStorySync)
        .catch((err: unknown) => {
          log("ERROR", `Storyjson sync threw unexpectedly: ${err instanceof Error ? err.message : String(err)}`);
          return true;
        })
        .then((result) => {
          if (result !== "done" && !forceCheck) {
            scheduleSyncRetry("Storyjson", runStorySync);
          }
        }),
    );
  } else {
    log("INFO", `STORYJSON_PATH is set (${process.env["STORYJSON_PATH"]}); story auto-sync disabled.`);
  }

  // Images artwork sync (2.5.0) — LOCAL_IMAGE mode consumes AKDP assets.
  if (cfg.imagesEnabled && cfg.localImage) {
    const runImagesSync = async (): Promise<boolean> => {
      const r = await syncImages(cfg.imagesPath, {
        includeOriginal: cfg.originalImage,
        forceCheck,
      });
      const sha = r.commitSha ? r.commitSha.slice(0, 8) : "unknown";
      if (r.status === "updated") {
        log("INFO", `Images updated from GitHub Release (${r.spec.repo} @ ${sha}).`);
      } else if (r.status === "up_to_date") {
        log("INFO", `Images are up to date (${r.spec.repo} @ ${sha}).`);
      } else if (r.status === "offline_fallback") {
        log("WARN", `Network unavailable; using cached images (${r.spec.repo} @ ${sha}). Error: ${r.error}`);
      } else {
        log("WARN", `Images sync failed — no data. Error: ${r.error}`);
      }
      return shouldRetrySync(r.status);
    };

    startupTasks.push(
      singleFlightSync("Images", runImagesSync)
        .catch((err: unknown) => {
          log("ERROR", `Images sync threw unexpectedly: ${err instanceof Error ? err.message : String(err)}`);
          return true;
        })
        .then((result) => {
          if (result !== "done" && !forceCheck) {
            scheduleSyncRetry("Images", runImagesSync);
          }
        }),
    );
  }

  await Promise.all(startupTasks);
}

export function resolveAutoSyncIntervalMs(
  raw = process.env["PRTS_AUTO_SYNC_INTERVAL_SECONDS"],
): number {
  if (raw === undefined) return AUTO_SYNC_DEFAULT_INTERVAL_SECONDS * 1000;
  const normalized = raw.trim();
  const interval = /^[+-]?\d+$/.test(normalized) ? Number(normalized) : Number.NaN;
  if (interval === 0) return 0;
  if (
    !Number.isInteger(interval)
    || interval < AUTO_SYNC_MIN_INTERVAL_SECONDS
    || interval > AUTO_SYNC_MAX_INTERVAL_SECONDS
  ) {
    log(
      "WARN",
      "Invalid PRTS_AUTO_SYNC_INTERVAL_SECONDS="
        + JSON.stringify(raw)
        + "; using "
        + AUTO_SYNC_DEFAULT_INTERVAL_SECONDS
        + "s.",
    );
    return AUTO_SYNC_DEFAULT_INTERVAL_SECONDS * 1000;
  }
  return interval * 1000;
}

interface AutoSyncTimer {
  unref?: () => void;
}

type AutoSyncSchedule = (
  callback: () => void,
  delayMs: number,
) => AutoSyncTimer;

export function runAutoSyncLoop(
  runSync: (forceCheck: boolean) => Promise<void> = runStartupSync,
  intervalMs = resolveAutoSyncIntervalMs(),
  schedule: AutoSyncSchedule = (callback, delayMs) =>
    setTimeout(callback, delayMs),
): void {
  let forceCheck = false;

  const runCycle = (): void => {
    void Promise.resolve()
      .then(() => runSync(forceCheck))
      .catch((err: unknown) => {
        const message = err instanceof Error ? err.message : String(err);
        log("ERROR", "Auto-sync cycle threw unexpectedly: " + message);
      })
      .finally(() => {
        if (intervalMs === 0) {
          log("INFO", "Periodic auto-sync is disabled; startup sync completed.");
          return;
        }
        log(
          "INFO",
          "Next auto-sync check in " + Math.round(intervalMs / 1000) + "s.",
        );
        const timer = schedule(() => {
          forceCheck = true;
          runCycle();
        }, intervalMs);
        timer.unref?.();
      });
  };

  runCycle();
}

export function startAutoSync(): void {
  if (autoSyncStarted) {
    log("INFO", "Auto-sync is already running; skipping duplicate start.");
    return;
  }
  autoSyncStarted = true;
  runAutoSyncLoop();
}

File: python/src/prts_mcp/startup_sync.py
"""Startup data sync orchestration and shared helpers.

Split from server.py. Owns the retry/backoff scheduling, single-flight
locking, and the cache-clearing cascade triggered when sync writes new data.
"""
from __future__ import annotations

import logging
import os
import threading
from pathlib import Path
from typing import Callable, Literal

_logger = logging.getLogger("prts_mcp.server")

_SYNC_RETRY_DELAYS_SECONDS = (30, 120, 600)
_AUTO_SYNC_DEFAULT_INTERVAL_SECONDS = 3600
_AUTO_SYNC_MIN_INTERVAL_SECONDS = 60
_AUTO_SYNC_MAX_INTERVAL_SECONDS = 7 * 24 * 60 * 60
# Startup sync labels are fixed domains; keep their locks for process lifetime
# so overlapping initial/retry attempts share the same mutex.
_SYNC_LOCKS: dict[str, threading.Lock] = {}
_SYNC_LOCKS_GUARD = threading.Lock()
_SyncRunResult = Literal["retry", "done", "skipped"]


def _require_story_zip(cfg) -> Path:  # type: ignore[no-untyped-def]
    """Return effective_storyjson_zip or raise RuntimeError.

    Takes a Config to avoid importing config at module load time (which would
    create a circular dependency during test stubbing).
    """
    if not cfg.has_story_data:
        raise RuntimeError(
            "剧情数据未就绪。请设置 STORYJSON_PATH 环境变量指向 zh_CN.zip，"
            "或等待服务器自动从 GitHub Release 下载完成后重试。"
        )
    return cfg.effective_storyjson_zip


def _sync_needs_retry(status: str) -> bool:
    return status in {"offline_fallback", "no_data"}


def _gamedata_pair_needs_retry(excel_status: str, levels_status: str) -> bool:
    """Decide whether the gamedata pair sync warrants a dense retry.

    A generation (commit_sha) mismatch alone must NOT trigger dense retries
    (30s/120s/600s): when both archives are up-to-date, divergent
    commit_shas are stale local metadata, not missing data, and the next
    periodic cycle resolves it. Only offline_fallback / no_data retry.
    """
    return _sync_needs_retry(excel_status) or _sync_needs_retry(levels_status)


def _run_initial_sync(label: str, sync_func: Callable[[], bool]) -> bool:
    """Run the first sync attempt, treating unexpected exceptions as retry-needed."""
    try:
        return sync_func()
    except Exception as exc:  # noqa: BLE001
        _logger.exception("%s sync threw unexpectedly: %s", label, exc)
        return True


def _single_flight_sync(label: str, sync_func: Callable[[], bool]) -> _SyncRunResult:
    with _SYNC_LOCKS_GUARD:
        lock = _SYNC_LOCKS.setdefault(label, threading.Lock())
    if not lock.acquire(blocking=False):
        _logger.info("%s sync is already running; skipping overlapping attempt.", label)
        return "skipped"
    try:
        return "retry" if sync_func() else "done"
    finally:
        lock.release()


def _schedule_sync_retry(label: str, sync_func: Callable[[], bool], attempt: int = 0) -> None:
    delay = _SYNC_RETRY_DELAYS_SECONDS[attempt] if attempt < len(_SYNC_RETRY_DELAYS_SECONDS) else None
    if delay is None:
        _logger.warning(
            "%s sync still needs retry after %s attempts; "
            "waiting for the next periodic cycle or process start.",
            label,
            len(_SYNC_RETRY_DELAYS_SECONDS),
        )
        return

    def _retry() -> None:
        try:
            result = _single_flight_sync(label, sync_func)
        except Exception as exc:  # noqa: BLE001
            _logger.exception("%s retry sync threw unexpectedly: %s", label, exc)
            result = "retry"
        if result == "skipped":
            _schedule_sync_retry(label, sync_func, attempt)
        elif result == "retry":
            _schedule_sync_retry(label, sync_func, attempt + 1)

    timer = threading.Timer(delay, _retry)
    timer.daemon = True
    timer.start()
    _logger.info("%s sync will retry in %ss.", label, delay)


def _log_sync_result(r) -> None:  # type: ignore[no-untyped-def]
    repo = r.spec.repo
    sha_short = r.commit_sha[:8] if r.commit_sha else "unknown"
    if r.status == "updated":
        _logger.info("Data updated from GitHub (%s @ %s).", repo, sha_short)
    elif r.status == "up_to_date":
        _logger.info("Data is up to date (%s @ %s).", repo, sha_short)
    elif r.status == "offline_fallback":
        _logger.warning(
            "Network unavailable; using cached data (%s @ %s). Error: %s",
            repo, sha_short, r.error,
        )
    elif r.status == "no_data":
        _logger.warning(
            "Sync failed for %s — no data available. Error: %s",
            repo, r.error,
        )


def _run_startup_sync(*, force_check: bool = False) -> None:
    """Check upstream GitHub and download data files if outdated.

    Skipped when GAMEDATA_PATH is explicitly set to a custom location —
    in that case the user is managing their own data and we must not
    overwrite it.
    """
    from prts_mcp.config import Config, DEFAULT_GAMEDATA_PATH
    from prts_mcp.data.datasets import GAMEDATA_EXCEL, GAMEDATA_LEVELS, STORY_ZH_CN
    from prts_mcp.data.sync import sync_release, sync_release_archive_pair

    cfg = Config.load()
    if cfg.is_custom_gamedata:
        _logger.info(
            "GAMEDATA_PATH is set to a custom location (%s); auto-sync disabled.",
            cfg.gamedata_path,
        )
    else:
        archive_spec = GAMEDATA_EXCEL.archive_spec(
            local_zip=DEFAULT_GAMEDATA_PATH / "archives" / "zh_CN-excel.zip",
            local_root=DEFAULT_GAMEDATA_PATH,
        )
        levels_spec = GAMEDATA_LEVELS.archive_spec(
            local_zip=cfg.levels_path / "archives" / "zh_CN-levels.zip",
            local_root=cfg.levels_path,
        )

        def _sync_gamedata_pair() -> bool:
            excel_result, levels_result = sync_release_archive_pair(
                archive_spec,
                levels_spec,
                force_check=force_check,
            )
            _log_sync_result(excel_result)
            _log_sync_result(levels_result)
            same_generation = (
                excel_result.commit_sha is not None
                and excel_result.commit_sha == levels_result.commit_sha
            )
            if same_generation and (
                excel_result.status == "updated"
                or levels_result.status == "updated"
            ):
                # Import for the registration side effect, then clear via the
                # dataset registry (single source for the gamedata domain list).
                from prts_mcp.data import (  # noqa: F401
                    building,
                    enemy,
                    item,
                    operator,
                    stage,
                    stage_enemy,
                )
                from prts_mcp.data.dataset_access import dataset_registry

                for domain in (
                    "operator", "enemy", "item", "stage_enemy", "stage", "building",
                ):
                    dataset_registry()[domain].clear()
            return _gamedata_pair_needs_retry(
                excel_result.status, levels_result.status
            )

        needs_retry = _run_initial_sync(
            "Gamedata",
            lambda: _single_flight_sync("Gamedata", _sync_gamedata_pair) != "done",
        )
        if needs_retry and not force_check:
            _schedule_sync_retry("Gamedata", _sync_gamedata_pair)

    # Always try to sync storyjson from GitHub Release (unless user supplied their own zip)
    if "STORYJSON_PATH" not in os.environ:
        release_spec = STORY_ZH_CN.release_spec(cfg.storyjson_zip)

        def _sync_storyjson() -> bool:
            r = sync_release(release_spec, force_check=force_check)
            _log_sync_result(r)
            if r.status == "updated":
                from prts_mcp.data.story import clear_story_caches

                clear_story_caches()
            return _sync_needs_retry(r.status)

        needs_retry = _run_initial_sync(
            "Storyjson",
            lambda: _single_flight_sync("Storyjson", _sync_storyjson) != "done",
        )
        if needs_retry and not force_check:
            _schedule_sync_retry("Storyjson", _sync_storyjson)

    # Images artwork sync (2.5.0) — LOCAL_IMAGE mode consumes AKDP assets.
    if cfg.images_enabled and cfg.local_image:
        from prts_mcp.sync.images_sync import sync_images as _sync_images_release

        def _sync_images() -> bool:
            r = _sync_images_release(
                cfg.images_path,
                include_original=cfg.original_image,
                force_check=force_check,
            )
            _log_sync_result(r)
            return _sync_needs_retry(r.status)

        needs_retry = _run_initial_sync(
            "Images",
            lambda: _single_flight_sync("Images", _sync_images) != "done",
        )
        if needs_retry and not force_check:
            _schedule_sync_retry("Images", _sync_images)


def _auto_sync_interval_seconds() -> int:
    raw = os.environ.get("PRTS_AUTO_SYNC_INTERVAL_SECONDS")
    if raw is None:
        return _AUTO_SYNC_DEFAULT_INTERVAL_SECONDS
    try:
        if not raw.strip():
            raise ValueError
        interval = int(raw)
    except ValueError:
        interval = -1
    if interval == 0:
        return 0
    if not (
        _AUTO_SYNC_MIN_INTERVAL_SECONDS
        <= interval
        <= _AUTO_SYNC_MAX_INTERVAL_SECONDS
    ):
        _logger.warning(
            "Invalid PRTS_AUTO_SYNC_INTERVAL_SECONDS=%r; using %ss.",
            raw,
            _AUTO_SYNC_DEFAULT_INTERVAL_SECONDS,
        )
        return _AUTO_SYNC_DEFAULT_INTERVAL_SECONDS
    return interval


def _run_auto_sync(stop_event: threading.Event | None = None) -> None:
    """Run startup sync, then keep checking Releases for the process lifetime."""
    stop = stop_event or threading.Event()
    interval = _auto_sync_interval_seconds()
    force_check = False

    while True:
        try:
            _run_startup_sync(force_check=force_check)
        except Exception as exc:  # noqa: BLE001
            _logger.exception("Auto-sync cycle threw unexpectedly: %s", exc)

        if interval == 0:
            _logger.info("Periodic auto-sync is disabled; startup sync completed.")
            return

        _logger.info("Next auto-sync check in %ss.", interval)
        if stop.wait(interval):
            return
        force_check = True
